rpcmiddleware/interface.go

```diff
@@ -0,0 +1,345 @@
+package rpcmiddleware
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+var (
+	// errorType is the reflection type of the error interface.
+	errorType = reflect.TypeOf((*error)(nil)).Elem()
+
+	// protoMessageType is the reflection type of the proto.Message
+	// interface.
+	protoMessageType = reflect.TypeOf((*proto.Message)(nil)).Elem()
+
+	// DefaultInterceptTimeout is the default maximum time we're allowed to
+	// take to respond to an RPC middleware interception request.
+	DefaultInterceptTimeout = 3 * time.Second
+
+	// acceptRequestHandler is a RequestHandler that accepts all requests.
+	acceptRequestHandler requestHandler = func(proto.Message) error {
+		return nil
+	}
+
+	// passThroughResponseHandler is a responseHandler that does not modify
+	// the response and just passes it through.
+	passThroughResponseHandler responseHandler = func(
+		proto.Message) (proto.Message, error) {
+
+		return nil, nil
+	}
+)
+
+// RequestInterceptor is a type that can intercept an RPC request.
+type RequestInterceptor interface {
+	// Name returns the name of the interceptor.
+	Name() string
+
+	// ReadOnly returns true if this interceptor should be registered in
+	// read-only mode. In read-only mode no custom caveat name can be
+	// specified.
+	ReadOnly() bool
+
+	// CustomCaveatName returns the name of the custom caveat that is
+	// expected to be handled by this interceptor. Cannot be specified in
+	// read-only mode.
+	CustomCaveatName() string
+
+	// Intercept processes an RPC middleware interception request and
+	// returns the interception result which either accepts or rejects the
+	// intercepted message.
+	Intercept(context.Context,
+		*lnrpc.RPCMiddlewareRequest) (*lnrpc.RPCMiddlewareResponse,
+		error)
+}
+
+// requestHandler is a function type for a generic gRPC request message handler
+// that can accept (=return nil) or refuse (=return non-nil error with rejection
+// reason) an incoming request.
+type requestHandler func(req proto.Message) error
```

Comment:
Should we also permit the request handler to modify the request (so return a new one), as we did in `lnd` recently?